src/Compilers/CSharp/Portable/Binder/Binder_Attributes.cs

```diff
@@ -121,14 +121,14 @@ internal partial class Binder
             BindingDiagnosticBag diagnostics)
         {
             beforeAttributePartBound?.Invoke(node);
-            var boundAttribute = new ExecutableCodeBinder(node, this.ContainingMemberOrLambda, this).BindAttribute(node, boundAttributeType, diagnostics);
+            var boundAttribute = new ExecutableCodeBinder(node, this.ContainingMemberOrLambda, this).BindAttribute(node, boundAttributeType, diagnostics, (this as ContextualAttributeBinder)?.AttributedMember);
             afterAttributePartBound?.Invoke(node);
             return (GetAttribute(boundAttribute, diagnostics), boundAttribute);
         }
 
-        internal BoundAttribute BindAttribute(AttributeSyntax node, NamedTypeSymbol attributeType, BindingDiagnosticBag diagnostics)
+        internal BoundAttribute BindAttribute(AttributeSyntax node, NamedTypeSymbol attributeType, BindingDiagnosticBag diagnostics, Symbol? attributedMember = null)
```

Comment:
Are we sure that callers not passing this parameter work properly? For example `CSharpSemanticModel.GetSpeculativeSymbolInfo(int position, AttributeSyntax attribute)` comes across this code path. This parameter as far as I understand affects `CallerMemberName` which I think there is no speculative semantic model tests for.

There is a second caller in `AttributeSemanticModel` that needs to be revised, I think.

Reply:
I wasn't able to come up with a speculative test which observes the value of implicit default arguments *before* this change. After this change, we pass the following test:

```cs
        [Fact]
        public void TestAttributeCallerInfoSemanticModel_Speculative()
        {
            var source = @"
using System;
using System.Runtime.CompilerServices;

class Attr : Attribute { public Attr([CallerMemberName] string s = null) { } }

class C
{
    [Attr(""a"")]
    void M0() { }
}
";
            var comp = CreateCompilation(source);
            comp.VerifyDiagnostics();

            var tree = comp.SyntaxTrees[0];
            var root = tree.GetRoot();
            var attrSyntax = root.DescendantNodes().OfType<AttributeSyntax>().Last();

            var semanticModel = comp.GetSemanticModel(tree);

            var newRoot = root.ReplaceNode(attrSyntax, attrSyntax.WithArgumentList(SyntaxFactory.ParseAttributeArgumentList("()")));
            var newAttrSyntax = newRoot.DescendantNodes().OfType<AttributeSyntax>().Last();

            Assert.True(semanticModel.TryGetSpeculativeSemanticModel(attrSyntax.ArgumentList.Position, newAttrSyntax, out var speculativeModel));

            var operation = speculativeModel.GetOperation(newAttrSyntax);
            VerifyOperationTree(comp, operation, @"
IOperation:  (OperationKind.None, Type: Attr) (Syntax: 'Attr()')
  Children(1):
      IDefaultValueOperation (OperationKind.DefaultValue, Type: System.String, Constant: null, IsImplicit) (Syntax: 'Attr()')");
        }
```

Before this PR, the 'None' operation has no children.
In other words, this implementation doesn't propagate the caller info properly. To do that, we would actually need to determine the attributedMember. Perhaps the implementation of `IAttributeOperation` would be a reasonable time to solve that.

Reply:
Getting `IOperartion` from speculative semantic model sounds interesting!

Reply:
FWIW, I think it could be difficult to do properly. I don't know how the rest of the team feels about it but it's the kind of thing I think would be reasonable to file as a follow-up bug after initial implementation of `IAttributeOperation`. It seems like the use case for CallerInfo is at runtime. If you are writing a tool that uses IOperation, etc. there's no reason to dig into the default arguments to get member name, line number, etc. values, just use the compiler APIs directly at that point.